graph/path/earliest_test.go

```diff
@@ -0,0 +1,160 @@
+	// graph I sketched in a notebook to demonstrate that
+	// my method for earliest arrival path construction
+	// worked OK.
+	sketchStream = []struct{
+		F, T int64
+		A, D uint64
+	}{
+		{1, 2, 1, 0},
+		{1, 3, 1, 0},
+		{4, 7, 1, 0},
+		{1, 4, 2, 0},
+		{2, 6, 3, 0},
+		{2, 5, 4, 0},
+		{5, 8, 5, 0},
+		{6, 8, 6, 0},
+		{7, 9, 7, 0},
+		{6, 9, 8, 0},
+	}
+
+	// wuChengHuangKeLuXuStream is the example temporal graph from
+	// figure 1(a) of https://www.vldb.org/pvldb/vol7/p721-wu.pdf
+	wuChengHuangKeLuXuStream = []struct{
+		F, T int64
+		A, D uint64
+	}{
+		{1, 2, 1, 0},
+		{1, 2, 2, 0},
+		{7, 10, 2, 0},
+		{1, 6, 3, 0},
+		{2, 7, 3, 0},
+		{2, 8, 3, 0},
+		{1, 3, 4, 0},
+		{6, 9, 5, 0},
+		{3, 8, 6, 0},
+		{7, 11, 6, 0},
+		{8, 9, 7, 0},
+		{9, 12, 8, 0},
+		{9, 12, 9, 0},
+		{1, 9, 10, 0},
+	}
+)
+
+func lineAt(fid, tid int64, a, d uint64) temporal.TemporalLine {
+	f := temporal.Node(fid)
+	t := temporal.Node(tid)
+	return temporal.TemporalLine{
+		F: &f,
+		T: &t,
+		A: a,
+		D: d,
+	}
+}
+
+func TestEarliestArrivalFrom(t *testing.T) {
+	tests := []struct {
+		name string
+		stream []struct{
+			F, T int64
+			A, D uint64
+		}
+		nodes map[int64]struct{
+			earliest uint64
+			path     []int64
+		}
+		from  int64
+		at    uint64
+		until uint64
+	}{
+		{
+			"sketchStream",
+			sketchStream,
+			map[int64]struct{
+				earliest uint64
+				path     []int64
+			}{
+				1: {0, []int64{1}},
+				2: {1, []int64{1,2}},
+				3: {1, []int64{1,3}},
+				4: {2, []int64{1,4}},
+				5: {4, []int64{1,2,5}},
+				6: {3, []int64{1,2,6}},
+				8: {5, []int64{1,2,5,8}},
+				9: {8, []int64{1,2,6,9}},
+			},
+			1,
+			0,
+			temporal.MaxTime,
+		},
+		{
+			"wuChengHuangKeLuXuStream",
+			wuChengHuangKeLuXuStream,
+			map[int64]struct{
+				earliest uint64
+				path     []int64
+			}{
+				1:  {0, []int64{1}},
+				2:  {1, []int64{1,2}},
+				3:  {4, []int64{1,3}},
+				6:  {3, []int64{1,6}},
+				7:  {3, []int64{1,2,7}},
+				8:  {3, []int64{1,2,8}},
+				9:  {5, []int64{1,6,9}},
+				11: {6, []int64{1,2,7,11}},
+				12: {8, []int64{1,6,9,12}},
+			},
+			1,
+			0,
+			temporal.MaxTime,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			lines := make([]graph.TemporalLine, len(test.stream))
+			for i, line := range test.stream {
+				lines[i] = lineAt(line.F, line.T, line.A, line.D)
+			}
+			it := iterator.NewOrderedTemporalLines(lines)
+			from := temporal.Node(test.from)
+			out := EarliestArrivalFrom(it, &from, test.at, test.until)
+			if out.Len() != len(test.nodes) {
+				t.Fatalf("unexpected reachable node: expected %v, got %v", len(test.nodes), out.Len())
+			}
+			for vid, ev := range test.nodes {
+				path, earliest := out.To(vid)
+				if path == nil {
+					t.Fatalf("expected reachable node: %v", vid)
+				}
+				if earliest != ev.earliest {
+					t.Fatalf("expected earliest arrival time: (%v) expected %v, got %v", vid, ev.earliest, earliest)
+				}
+				if len(path) != len(ev.path) {
```

Comment:
Use `reflect.DeepEqual` instead of `goto`.

Reply:
`reflect.DeepEqual` does not work. (array of interfaces vs. array of values)

Reply:
Either make `ev.path` a `[]simple.Node` or make a helper to give you a slice of IDs from a `[]graph.Node`.

I see that you retain only int64 in the `Earlier`. We don't do this. We keep the nodes since they may be used by the client for annotation.